File: 3016-minimum-number-of-pushes-to-type-word-ii/3016-minimum-number-of-pushes-to-type-word-ii.cpp
class Solution {
public:
    int minimumPushes(string word) {
        // Frequency vector to store count of each letter
        vector<int> frequency(26, 0);

        // Count occurrences of each letter
        for (char& c : word) {
            ++frequency[c - 'a'];
        }

        // Sort frequencies in descending order
        sort(frequency.rbegin(), frequency.rend());

        int totalPushes = 0;

        // Calculate total number of presses
        for (int i = 0; i < 26; ++i) {
            if (frequency[i] == 0) break;
            totalPushes += (i / 8 + 1) * frequency[i];
        }

        return totalPushes;
    }
};

File: 3016-minimum-number-of-pushes-to-type-word-ii/3016-minimum-number-of-pushes-to-type-word-ii.js
/**
 * @param {string} word
 * @return {number}
 */
var minimumPushes = function(word) {

    const freq = Array(26).fill(0);

    for (const c of word)
        freq[c.charCodeAt(0) - 97]++;

    freq.sort((a, b) => b - a);

    let ans = 0;

    for (let i = 0; i < 26 && freq[i] > 0; i++) {

        ans += freq[i] * (Math.floor(i / 8) + 1);
    }

    return ans;
};